dhall/src/Dhall.hs

```diff
@@ -791,6 +800,126 @@ list = fmap Data.Foldable.toList . sequence
+
+Duplicate elements are ignored.
+
+>>> input (setIgnoringDuplicates natural) "[1, 1, 3]"
+fromList [1,3]
+
+-}
+setIgnoringDuplicates :: (Ord a) => Type a -> Type (Data.Set.Set a)
+setIgnoringDuplicates = fmap Data.Set.fromList . list
+
+{-| Decode a `HashSet` from a `List`
+
+>>> input (hashSetIgnoringDuplicates natural) "[1, 2, 3]"
+fromList [1,2,3]
+
+Duplicate elements are ignored.
+
+>>> input (hashSetIgnoringDuplicates natural) "[1, 1, 3]"
+fromList [1,3]
+
+-}
+hashSetIgnoringDuplicates :: (Hashable a, Ord a)
+                          => Type a
+                          -> Type (Data.HashSet.HashSet a)
+hashSetIgnoringDuplicates = fmap Data.HashSet.fromList . list
+
+{-| Decode a `Set` from a `List` with distinct elements
+
+>>> input (setFromDistinctList natural) "[1, 2, 3]"
+fromList [1,2,3]
+
+An error is thrown if the list contains duplicates.
+
+> >>> input (setFromDistinctList natural) "[1, 1, 3]"
+> *** Exception: Error: Failed extraction
+>
+> The expression type-checked successfully but the transformation to the target
+> type failed with the following error:
+>
+> One duplicate element in the list: 1
+>
+
+> >>> input (setFromDistinctList natural) "[1, 1, 3, 3]"
+> *** Exception: Error: Failed extraction
+>
+> The expression type-checked successfully but the transformation to the target
+> type failed with the following error:
+>
+> 2 duplicates were found in the list, including 1
+>
+
+-}
+setFromDistinctList :: (Ord a, Show a) => Type a -> Type (Data.Set.Set a)
+setFromDistinctList = setHelper Data.Set.size Data.Set.fromList
+
+{-| Decode a `HashSet` from a `List` with distinct elements
+
+>>> input (hashSetFromDistinctList natural) "[1, 2, 3]"
+fromList [1,2,3]
+
+An error is thrown if the list contains duplicates.
+
+> >>> input (hashSetFromDistinctList natural) "[1, 1, 3]"
+> *** Exception: Error: Failed extraction
+>
+> The expression type-checked successfully but the transformation to the target
+> type failed with the following error:
+>
+> One duplicate element in the list: 1
+>
+
+> >>> input (hashSetFromDistinctList natural) "[1, 1, 3, 3]"
+> *** Exception: Error: Failed extraction
+>
+> The expression type-checked successfully but the transformation to the target
+> type failed with the following error:
+>
+> 2 duplicates were found in the list, including 1
+>
+
+-}
+hashSetFromDistinctList :: (Hashable a, Ord a, Show a)
+                        => Type a
+                        -> Type (Data.HashSet.HashSet a)
+hashSetFromDistinctList = setHelper Data.HashSet.size Data.HashSet.fromList
+
+
+setHelper :: (Eq a, Foldable t, Show a)
+          => (t a -> Int)
+          -> ([a] -> t a)
+          -> Type a
+          -> Type (t a)
+setHelper size toSet (Type extractIn expectedIn) = Type extractOut expectedOut
+  where
+    extractOut (ListLit _ es) = case traverse extractIn es of
+        Success vSeq
+            | sameSize               -> Success vSet
+            | length duplicates == 1 -> extractError err1
```

Comment:
I think something like this would make it cheaper to distinguish `Success` and `Failure`:

```suggestion
            | otherwise -> extractError err
          where err | length duplicates == 1 …
```

Reply:
Why would it be cheaper?
I don't mind, but `err1` and `errN` take up a few lines, I think the structure code may be harder to read...

Reply:
In the current situation it seems that we need to compute `length duplicates` to get the `Failure` constructor. I believe that if we move the `length duplicates` check into the computation of the error message, the caller can determine the constructor without computing `length duplicates`.

Maybe `GHC` is smart enough, but it's better not to rely on it too much.

Reply:
But `Failure` is only accessed if it's not a `Success`. If it's not a success, it doesn't go in the conditions.

```haskell
f x = case x of
    Just x | error "whoops" -> 1
    Nothing -> 0

ghci> f Nothing
0
```

Reply:
I'm not quite sure what you mean. `extractError` results in `Failure` too.

I'm assuming usage like

```
case toMonadic (extract (setFromDistinctList elementType) expr) of
  Right s -> …
  Left _ -> … -- handle failure but ignore the error message
```

I'm arguing that even if `expr` contains duplicates, it shouldn't be necessary to force `length duplicates`, but the current code looks like it does.

Reply:
OK, I understand what you mean now, users should be able to pattern match on `Failure` without extra computation. That's a good point and sorry for being so dense ^^

Reply:
No worries. I wasn't very clear!